csrc/multidevice/utils.h

```diff
@@ -127,4 +129,44 @@ int64_t getShardedAxis(TensorView*);
 
 // Reorders a TensorView so that the DID parallelized axis are in front.
 void reorderDIDToFront(TensorView*);
+
+// Given a TensorView and the shape of a sharded tensor of which certain
+// dimensions are partially alloated, returns the global shape that'll be used
+// to bind to the TensorView's logical domain. This is to solve #3282 so we can
+// bind a sharded tensor to a TensorView that has a DID-parallel loop domain.
+//
+// For example, when `tv` is
+//   logical: iM, iN
+//   allocation: iDIDx{D}, iN/D, iM
+// and `sizes` is [2, 3], the returned shape will be [2, 3D]. This is because,
```

Comment:
I think there is a mistake here..
if we bind `{2,3}` to `{N/D, M}`
then `M=3` and `N=2D`, and so according to the comment, it should return the shape corresponding to the logical domain, i.e., `[3, 2D]`. Am I missing something?

Reply:
moreover, do we support transposition?

Reply:
> The comment is correct and is consistent with the code.

I see. Imo it is error prone to silently discard transposition. We should assert that only splits have been applied, OR, we should support transposition, which shouldn't be too hard...



> ExpressionEvaluator::bindTensorDomain **basically** does the following
> 
> ```
> unsharded_sizes = unshardedSizes(t.sizes());
> for (i : range(t.dim())) {
>   bind(logical_domain[i], unsharded_sizes[i]);
> }
> ```
> 
> That's also why I prefer to say we bind the unsharded sizes to the logical domain instead of allocation.


I would say in this case that we bind to neither the logical nor the allocation, but to some hybrid domain where starting from the logical we only applied the splits. This is a bit counter-intuitive to me.

In your snippet above, everything is contained in the `unsharded_sizes` which basically embeds a mapping from allocation (or more precisely the hybrid domain I described earlier) to logical.

Reply:
> it is error prone to silently discard transposition.

I believe code as is supports transposition. (I assume by transposition you mean TensorView::reorder). To assure you that, I added a test in the latest commit.